cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+This design has a few wins here:
+- The Authentication only requires that the TLS Handshake succeed, it does not require that we keep the connection open, so we will be making short lived connections for each authentication check, and this will most likely need to be to a new, dedicated port.
+- Because we will be controlling both ends of the TLS handshake, we can use any field we like in the certificates for authentication - SPIFFE certificates issued by SPIRE have the full SPIFFE ID URI as a SAN field, so we can use that for choosing both client and server certificates - the TLS request from the client will be requesting the full SPIFFE ID as its URI.
+- Because of the preceding points, having this port open is only risking buffer overflows etc, not incorrectly routing to some service. Additionally, because the SPIFFE IDs in use will only have the trust domain and Cilium Identity number in the URI, there's not much information leakage possible.
+
+
+#### Local agent Events
+
+##### Startup
+* If mesh-auth-mode is enabled, listen on mesh-auth-port for connections
+  * When accepting new connections, check the target SPIFFE ID URI and use GetCertificate() to locate the certificate corresponding to the local peer for the data session.
+  * Q: How does this server know which peer (i.e client pod) is initiating this connection? The exact Pod IP on the source node is actually not important, as long as we can trust the Cilium agent to look up the Cilium Identity correctly for any Pod running on it, then we only need the source Cilium Identity (which can be used to derive the certificate for that workload from the certificates the agent has already seen). In the SPIFFE case, the client certificate issued to the workload by the SPIRE server includes the SPIFFE ID as a SAN field of type URI, so the server knows which _identity_ is attempting to authenticate, and can validate the connection on that basis. Other future certificate stores will also need to ensure that a SPIFFE ID or equivalent is populated into a SAN in the certificate.
+  * Q: How to make missing certificate errors super obvious for debugging? We're attempting to route around this problem by having Pod startup blocked until the SPIRE server has completed issuing certificates for the identity given to any specific Pod, the same as we already do for Identity creation.
+
+##### CiliumNetworkPolicy creation
+Parse CNPs for the new authentication field. When this is present, as part of policy plumbing, set a new field in the policymap entries that indicates that authentication is required for connectivity to this peer.
```

Comment:
we should also take the config flag `mesh-auth-mode` into account when writing the policy map entry. if disabled, we should overrule the config from the policy.